cotel/src/main/scala/day3/SpiralMatrix.scala

```diff
@@ -0,0 +1,147 @@
+
+    for (x <- matrix.indices;
+         y <- matrix(x).indices) yield {
+      if (matrix(x)(y) == number) return Option((x, y))
+    }
+
+    None
+  }
+
+  private def calculateArraySize(i: Int, acc: Int = 0): Int = Math.pow(2 * acc + 1, 2).toInt match {
+    case corner if corner >= i => Math.sqrt(corner).toInt
+    case corner if corner < i => calculateArraySize(i, acc + 1)
+  }
+
+  private def generateSpiralMatrix(upToN: Int): Matrix[Int] = {
+    var fn = 1
+    var expo = 1
+
+    val arraySize = calculateArraySize(upToN)
+    var x = arraySize / 2
+    var y = arraySize / 2
+
+    val matrix: Matrix[Int] = Array.fill(arraySize)(Array.fill(arraySize)(0))
+
+    while (fn < Math.pow(arraySize, 2)) {
+      val per = Math.pow(expo, 2) - Math.pow(expo - 2, 2)
+      if (per == 0) matrix(x)(y) = 1
+      else {
+        x += 1
+        fn += 1
+        matrix(y)(x) = fn
+
+        for (_ <- 0 until expo - 2) yield {
+          y -= 1
+          fn += 1
+          matrix(y)(x) = fn
+        }
+
+        for (_ <- 1 until expo) yield {
+          x -= 1
+          fn += 1
+          matrix(y)(x) = fn
+        }
+
+        for (_ <- 1 until expo) yield {
+          y += 1
+          fn += 1
+          matrix(y)(x) = fn
+        }
+
+        for (_ <- 1 until expo) yield {
+          x += 1
+          fn += 1
+          matrix(y)(x) = fn
+        }
+      }
+
+      expo += 2
+    }
+
+    matrix
+  }
+
+  def firstNumberBiggerThanInput(upToN: Int): Int = {
+    var fn = 1
+    var expo = 1
+
+    val arraySize = calculateArraySize(upToN) + 2
+    var x = arraySize / 2
+    var y = arraySize / 2
+
+    val matrix: Matrix[Int] = Array.fill(arraySize)(Array.fill(arraySize)(0))
+
+    while (fn < Math.pow(arraySize, 2)) {
+      val per = Math.pow(expo, 2) - Math.pow(expo - 2, 2)
+      if (per == 0) matrix(x)(y) = 1
+      else {
+        x += 1
+        fn += 1
+        matrix(y)(x) = sumNeighbors(matrix, (x, y))
+        if (matrix(y)(x) > upToN) return matrix(y)(x)
+
+        for (_ <- 0 until expo - 2) yield {
+          y -= 1
+          fn += 1
+          matrix(y)(x) = sumNeighbors(matrix, (x, y))
+          if (matrix(y)(x) > upToN) return matrix(y)(x)
+        }
+
+        for (_ <- 1 until expo) yield {
+          x -= 1
+          fn += 1
+          matrix(y)(x) = sumNeighbors(matrix, (x, y))
+          if (matrix(y)(x) > upToN) return matrix(y)(x)
+        }
+
+        for (_ <- 1 until expo) yield {
+          y += 1
+          fn += 1
+          matrix(y)(x) = sumNeighbors(matrix, (x, y))
+          if (matrix(y)(x) > upToN) return matrix(y)(x)
+        }
+
+        for (_ <- 1 until expo) yield {
+          x += 1
+          fn += 1
+          matrix(y)(x) = sumNeighbors(matrix, (x, y))
+          if (matrix(y)(x) > upToN) return matrix(y)(x)
+        }
+      }
+
+      expo += 2
+    }
+
+    -1
+  }
+
+  private def sumNeighbors(matrix: Matrix[Int], position: (Int, Int)): Int =
```

Comment:
How do you iterate over the matrix?

for i and j it's like a for inside for?

Reply:
Yes